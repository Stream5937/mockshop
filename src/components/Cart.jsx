import { Link, useOutletContext } from "react-router-dom";
import style from "../styles/cart.module.css";

export default function Cart() {
  const context = useOutletContext();
  const { products, error, loading } = context[0];
  const [cart, setCart] = context[1];
  console.log("cart: ", cart);
  const [removeFromCart, clearCart, sumCart] = context[2];
  const [quantity, setQuantity] = context[3];
  const [totalItems, setTotalItems] = context[4];

  return (
    <>
      <div className={style.cartContainer}>
        {/* Mini Cart */}

        <h2>ðŸ›’ Mini Cart (total items {totalItems})</h2>
        <div id={style.viewFull}>
          {cart.length > 0 && (
            <Link
              to="/cartPage"
              style={{ display: "block", marginTop: "0.5rem" }}
            >
              View Full Cart Page
            </Link>
          )}
        </div>

        {cart.length === 0 ? (
          <p>No items yet</p>
        ) : (
          <ul
            id={style.cart_ul}
            style={{ listStyle: "none", paddingLeft: 0, margin: 0 }}
          >
            {cart.map((obj) => (
              <li key={obj.item.id} style={{ marginBottom: "0.5rem" }}>
                <img
                  src={obj.item.image}
                  alt={obj.item.title}
                  style={{
                    verticalAlign: "middle",
                    objectFit: "contain",
                  }}
                />
                <span id={style.text} style={{ marginLeft: "0.5rem" }}>
                  {obj.item.title.substring(0, 20)}...
                  <span id={style.num}>(x{obj.number})</span>
                </span>
                <button
                  id={style.li_btn}
                  value={obj.item.id}
                  onClick={() => removeFromCart(obj.item.id)}
                  style={{ marginLeft: "0.5rem" }}
                >
                  âœ–
                </button>
              </li>
            ))}
          </ul>
        )}
      </div>
    </>
  );
}
