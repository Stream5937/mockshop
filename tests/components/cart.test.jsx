import { describe, it, expect, vi, beforeEach } from "vitest";
import { page, screen } from "@testing-library/react";
import { renderWithRouter } from "../utils/renderWithRouter.jsx";
import Cart from "../../src/components/Cart.jsx";

vi.mock("react-router-dom", async () => {
  //mock totals
  let mockQuantity = 3;
  let mockTotalItems = 3;
  let mockTotalCost = 30;

  const mod = await vi.importActual("react-router-dom");

  const mockProduct = {
    id: 1,
    title: "Test Product",
    category: "Test Category",
    price: 10,
    description: "This is a test product description",
    rating: { rate: 4.5, count: 20 },
    image: "test.jpg",
    the_alt_text: "Test Product",
  };

  const mockContext = [
    { products: [mockProduct], error: null, loading: false }, // context[0]
    [[{ item: mockProduct, number: 1 }], vi.fn()], // context[1]
    [vi.fn(), vi.fn(), vi.fn()], // context[2]
    [mockQuantity, vi.fn()], // context[3]
    [mockTotalItems, vi.fn()], //context[4];
    [mockTotalCost, vi.fn()], //context[5]
  ];

  return {
    ...mod,
    useOutletContext: () => mockContext, // return array
  };
});

describe("Cart component", () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  it("renders", () => {
    const { asFragment } = renderWithRouter(<Cart />);
    expect(asFragment()).toMatchSnapshot();
  });

  /*
  //single item cart
  it("single product cart", () => {
    const testid = renderWithRouter(<Cart />);
    expect(testid("testImg").toBeInTheDocument);
    //expect(screen.getByRole("img").name.toMatch("test.jpg"));
  });
  */

  it("uses correct img src", async () => {
    const { getByAltText } = await renderWithRouter(<Cart />);

    const image = getByAltText("Test Product");

    // expect(image.src).toContain("the_url");
    // or
    expect(image).toHaveAttribute("src", "test.jpg");
  });

  //cart total items
  it("with three item cart: renders text", () => {
    renderWithRouter(<Cart />);
    expect(screen.getByRole("heading").textContent).toMatch(
      "ðŸ›’ Mini Cart (total items 3)"
    );
    expect(screen.getByText("Current Total to Pay: $30"));
  });
});
