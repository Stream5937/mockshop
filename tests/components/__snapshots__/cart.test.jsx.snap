// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Cart component > renders 1`] = `
<DocumentFragment>
  <div
    class="_cartContainer_6ddd39"
  >
    <h2>
      ðŸ›’ Mini Cart (total items 3)
    </h2>
    <div
      id="_viewFull_6ddd39"
    >
      <a
        data-discover="true"
        href="/cartPage"
        style="display: block; margin-top: 0.5rem;"
      >
        View Full Cart Page
      </a>
    </div>
    <ul
      id="_cart_ul_6ddd39"
      style="list-style: none; padding-left: 0px; margin: 0px;"
    >
      <li
        style="margin-bottom: 0.5rem;"
      >
        <img
          alt="Test Product"
          src="test.jpg"
          style="vertical-align: middle; object-fit: contain;"
        />
        <span
          id="_text_6ddd39"
          style="margin-left: 0.5rem;"
        >
          Test Product...
          <span
            id="_num_6ddd39"
          >
            (x1)
          </span>
        </span>
        <button
          id="_li_btn_6ddd39"
          style="margin-left: 0.5rem;"
          value="1"
        >
          âœ–
        </button>
      </li>
    </ul>
    <div
      id="_cartTotal_6ddd39"
    >
      <p>
        Current Total to Pay: $30 
      </p>
    </div>
  </div>
</DocumentFragment>
`;
